flang/lib/Lower/IntrinsicCall.cpp

```diff
@@ -664,12 +699,96 @@ static std::string getIntrinsicWrapperName(const llvm::StringRef &intrinsic,
   return name;
 }
 
+/// fir::ExtendedValue to mlir::Value translation layer
+
+fir::ExtendedValue toExtendedValue(mlir::Value val,
+                                   Fortran::lower::FirOpBuilder &builder,
+                                   mlir::Location loc) {
+  assert(val && "optional unhandled here");
+  auto type = val.getType();
+  auto base = val;
+  mlir::Value len;
+  auto indexType = builder.getIndexType();
+  llvm::SmallVector<mlir::Value, 2> extents;
+
+  if (auto refType = type.dyn_cast<fir::ReferenceType>())
+    type = refType.getEleTy();
+
+  if (auto arrayType = type.dyn_cast<fir::SequenceType>()) {
+    type = arrayType.getEleTy();
+    llvm::ArrayRef<fir::SequenceType::Extent> shape = arrayType.getShape();
+    if (auto charType = type.dyn_cast<fir::CharacterType>()) {
+      auto cstLen = arrayType.getShape()[0];
+      shape = shape.drop_front();
+      if (cstLen == fir::SequenceType::getUnknownExtent())
+        mlir::emitError(loc, "No runtime length for given mlir::Value");
+      len = builder.createIntegerConstant(loc, indexType, cstLen);
+    }
+    for (auto extent : shape) {
+      if (extent == fir::SequenceType::getUnknownExtent())
+        extents.emplace_back(mlir::Value{});
```

Comment:
I think this might be a problem as subsequent code won't expect a nullptr.

For an array value, we should always have a grip on its size, even if the size is dynamic. So, even if we have an array like
```fortran
T :: arr(:,:)
```
we'll be able to get its shape from the box, etc.

The point of `fir::ExtendedValue` is that tracking these runtime values will be needed, etc. Say there is a path where the array is reshaped.

Reply:
I had assumed size array in mind ? How should we create the `ExtendedValue for such case ? In that case we will never get the a dynamic value for last extent.
` T :: arr(10, *)`

Reply:
IIRC, we just don't pass anything for the final extent. It's not terribly useful unless we want to generate bounds checking code.

Reply:
OK, thanks for the explanation. I modified the code to do nothing if the last extent is `?` and to emit an error if other extents are missing (in which case a box should have be used in the wrapper interface).